### PR TITLE
Visualization templates backend

### LINKED ISSUE #4805: Visualization templates backend
- [x] DB model for Template
```
title text not null
description text not null default  ''
min_supported_version text null
max_supported_version text null
required_tables [uuid]
source_visualization_id uuid not null
code string not null default ''
organization_id uuid not null
```
- [x] ActiveRecord model
- [x] Specific logic to check that list of related tables have always format `schema.table_name`
- [x] Rails endpoint with `create`, `index`, `update`, `delete`, `show`
- [x] `Related templates` functionality, so when you try to delete a visualization informs if there are dependant templates
- [x] Delete templates if visualization gets deleted
- [x] Delete templates if organization gets deleted
- [x] AR model specs
- [x] Rails endpoint tests
- [x] Make versions not mandatory
- [x] Endpoints for checking if a vis or table relates with 1+ template
- [x] Delete templates upon related models deletions (user, org, table)
(lost some GH points it looks to have reverted to an older version)

Use existing `templated_workflows` feature flag

Branch: `4805-tmplated-vis-backend`

See [Visualization-templates-technical-approach](https://github.com/CartoDB/cartodb-management/wiki/Visualization-templates-technical-approach) for internals

### PR DESCRIPTION
**Pull request by Kartones**
Closes #4805

### CODE REVIEW

**Comment by Kartones** [comment]
Created earlier to check how tests go with the downgrade of `activerecord-postgres-array` gem

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5525/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5528/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5533/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5537/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5541/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5542/))

**Comment by Kartones** [comment]
retest this please

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5553/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5561/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5566/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5567/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5574/))

**Comment by Kartones** [comment]
@juanignaciosl CR please

NOTE: On purpose the feat. flag is not checked on the vis/table controllers new endpoints because no matter if the user triggering the action has the flag or not, we should try to keep data integrity (as else we could be removing stuff without the user knowing).

@xavijam @javisantana Anything you miss or looks wrong?

Being feature flag based I'll release as soon as tests pass so you can use ir whenever you want.

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5590/))

**Comment by Kartones** [comment]
retest this please

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5595/))

**Comment by Kartones** [comment]
retest this please

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5599/))

**Comment by Kartones** [comment]
Last commit only removes the if inside template model because now there is no ambiguation as the only code fragment calling it is always from a `Table` model

File: app/controllers/carto/api/tables_controller.rb [hunk 1]
```diff
@@ -17,6 +17,14 @@ def show
         render_jsonp(table.public_values({request:request}, current_user).merge(schema: table.schema(reload: true)))
       end
 
+      def related_templates
+        templates = table.related_templates
+
+        render_jsonp({ items: templates.map { |template| Carto::Api::TemplatePresenter.new(template).public_values } })
+      rescue => e
+        render json: { error: [e.message] }, status: 400
```
Comment (juanignaciosl) [inline app/controllers/carto/api/tables_controller.rb]: This should be logged including `table` information.

Comment (Kartones) [inline app/controllers/carto/api/tables_controller.rb]: true, I forgot about logging errors... going to add it to all controller calls

File: app/controllers/carto/api/tables_controller.rb [hunk 2]
```diff
@@ -17,6 +17,14 @@ def show
         render_jsonp(table.public_values({request:request}, current_user).merge(schema: table.schema(reload: true)))
       end
 
+      def related_templates
```
Comment (juanignaciosl) [inline app/controllers/carto/api/tables_controller.rb]: Doesn't this need this `ssl_allowed` or `ssl_required`? It should've failed in staging :-\

File: app/controllers/carto/api/template_presenter.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+module Carto
+  module Api
+    class TemplatePresenter
+
+      attr_reader :template
```
Comment (juanignaciosl) [inline app/controllers/carto/api/template_presenter.rb]: Why exposing the injected `template`?

Comment (Kartones) [inline app/controllers/carto/api/template_presenter.rb]: bad habit, you're right as in the end I haven't used it, so taking it out

File: app/controllers/carto/api/template_presenter.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        @template = template
+      end
+
+      # When used inside a list, parent will be a " items: [ ... ] "
+      def public_values
+        {
+          id:                     @template.id,
+          title:                  @template.title,
+          description:            @template.description,
+          source_visualization:   {
```
Comment (juanignaciosl) [inline app/controllers/carto/api/template_presenter.rb]: Is this structure needed? If frontend only needs the `id` it'd be better sending `source_visualization_id`, which is clear. If you return a hash frontend might expect more attributes.

File: app/controllers/carto/api/tables_controller.rb (hunk 2 above)
Comment (Kartones) [inline app/controllers/carto/api/tables_controller.rb]: yup, I also noticed it while adding right now the error logging

File: app/controllers/carto/api/templates_controller.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+module Carto
+  module Api
+    class TemplatesController < ::Api::ApplicationController
+
+      ssl_required :index, :show, :create, :update, :destroy
+      before_filter :load_template, only: [ :show, :update, :destroy ]
+      before_filter :check_feature_flag
+
+      def index
+        templates = Carto::Template.where(organization_id: current_user.organization_id).order(:created_at).reverse
```
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: I think you actually want `reverse_order`. `reverse` is in-memory.

File: app/controllers/carto/api/template_presenter.rb (hunk 4 above)
Comment (Kartones) [inline app/controllers/carto/api/template_presenter.rb]: Yes, this was an old talk with @javisantana that, in the shake of consistency returning json data, we should always provide entity skeletons in the form `entity { id ... }`, so fetching is always the same `data[item]['id']` and then if has additional fields will read them, else only grabs the id, but at least we don't need to fiddle with having sometimes an `[entity][id]` and other times a `whatever_id`.

We didn't used it in the past because we said "only for new endpoints" but I recalled it and started to apply now.

/cc @rafatower @ethervoid

Comment (javisantana) [inline app/controllers/carto/api/template_presenter.rb]: https://github.com/interagent/http-api-design#nest-foreign-key-relations

that guide from heroku is gold

File: app/controllers/carto/api/templates_controller.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        render_jsonp({ items: templates.map { |template| Carto::Api::TemplatePresenter.new(template).public_values } })
+      rescue => e
+        render json: { error: [e.message] }, status: 400
+      end
+
+      def show
+        render_jsonp({ :errors => ["Template #{params[:id]} not found"] }, 404) and return if @template.nil?
```
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: Since you can configure which actions do load the template you could return this error in the `before_filter`, removing this simple checks from controller actions ([example](https://github.com/CartoDB/cartodb/blob/master/app/controllers/carto/api/visualizations_controller.rb#L143)).

File: app/controllers/carto/api/templates_controller.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+        render_jsonp(Carto::Api::TemplatePresenter.new(@template).public_values)
+      rescue => e
+        render json: { error: [e.message] }, status: 400
```
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: Same concern about login (I won't add this comment to remaining `rescue`s).

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5601/))

File: app/controllers/carto/api/templates_controller.rb [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      end
+
+      def update
+        render_jsonp({ :errors => ["Template #{params[:id]} not found"] }, 404) and return if @template.nil?
+
+        @template.title =                 params['title']
```
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: Not too relevant, but avoiding duplication between `create` and `update` would be nice.

File: app/controllers/carto/api/templates_controller.rb (hunk 5 above)
Comment (Kartones) [inline app/controllers/carto/api/templates_controller.rb]: thanks! I used the one from VisualizationQueryBuilder, so my bad for not checking why already was giving me items without a `.all`

File: app/controllers/carto/api/templates_controller.rb [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      rescue => e
+        render json: { error: [e.message] }, status: 400
+      end
+
+      private
+
+      def load_template
+        @template = Carto::Template.where(id: params[:id]).first
+      rescue
+        @template = nil
```
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: I'll comment this because it's a filter, not an action, I think this should be sent as an error.

File: app/controllers/carto/api/templates_controller.rb (hunk 8 above)
Comment (Kartones) [inline app/controllers/carto/api/templates_controller.rb]: Uhmm, I wouldn't do that (extracting to a method to reuse or similar), it is similar but not the same as some fields aren't touched, with the `before_filter` cleanings it will just be 8 lines and I'd rather have them and simple code and not an "update/create" method with ifs or something like that.

If it was more complex of course, is just that is very simple code and I don't see any benefit for just a few lines

File: app/controllers/carto/api/templates_controller.rb [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      end
+
+      def check_feature_flag
+        raise "Endpoint disabled for this user" unless current_user.has_feature_flag?('templated_workflows')
```
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: What about a 400 instead?

File: app/controllers/carto/api/visualizations_controller.rb [hunk 11]
```diff
@@ -115,6 +115,19 @@ def static_map
         redirect_to Carto::StaticMapsURLHelper.new.url_for_static_map(request, @visualization, map_width, map_height)
       end
 
+      def related_templates
```
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: 1.- This is probably missing `ssl_required` as well.
2.- Why not trusting `load_visualization` filter?
3.- Shouldn't this `related_templates` endpoint should only be at `TemplatesController`? You can point the `visualizations/:id/related_templates` endpoint and others to it, there's no constraint forcing that route to point to `VisualizationsController`.

File: app/controllers/carto/api/templates_controller.rb (hunk 10 above)
Comment (Kartones) [inline app/controllers/carto/api/templates_controller.rb]: Ok but a 403/forbidden then, as makes more sense than a 400

File: app/models/carto/template.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+    validates :organization_id, presence: true
+    validates :source_visualization_id, presence: true
+    validates :title, presence: true
+
+    validate :source_visualization_referential_integrity
+    validate :required_tables_should_be_qualified
+    validate :required_tables_referential_integrity
+
+    def ==(other_template)
```
Comment (juanignaciosl) [inline app/models/carto/template.rb]: Uhm, first time I see this in our models, why is it needed?

File: app/controllers/carto/api/templates_controller.rb (hunk 9 above)
Comment (Kartones) [inline app/controllers/carto/api/templates_controller.rb]: as it would fail when called I wasn't giving it much interest, but maybe yes, should directly be an error here instead of bubbling up

File: app/models/carto/template.rb [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      self.id == other_template.id
+    end
+
+    def required_tables
```
Comment (juanignaciosl) [inline app/models/carto/template.rb]: :( This is needed because of extension bug, isn't it? :( Document it, please.

File: app/models/carto/template.rb [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      self.required_tables_list.split(',')
+    end
+
+    def required_tables=(list=[])
+      self.required_tables_list = list.join(',')
+    end
+
+    def relates_to_table?(table)
+      table_name = "#{table.owner.database_schema}.#{table.name}"
```
Comment (juanignaciosl) [inline app/models/carto/template.rb]: IMHO this `table_name` belongs to `table`.

File: app/models/carto/template.rb (hunk 12 above)
Comment (Kartones) [inline app/models/carto/template.rb]: what is new? This are normal AR validations, plus some custom ones I built, similar to what we have on Sequel but adapted... is that what you mean?

File: app/models/carto/template.rb [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+      required_tables.include?(table_name)
+    end
+
+    private
+
+    def source_visualization_referential_integrity
```
Comment (juanignaciosl) [inline app/models/carto/template.rb]: Why is this needed? :-\

File: app/models/carto/template.rb [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+      errors.add(:source_visualization_id, "Source visualization not found") if visualization.nil?
+
+      if visualization.user.organization_id != self.organization_id
+        errors.add(:source_visualization_id, "Source visualization not found")
+      end
+    end
+
+    def required_tables_should_be_qualified
+      wrong_table_names = required_tables.select { |table_name|
+          (table_name =~ /^[a-z\-_0-9]+\.[a-z\-_0-9]+?$/) != 0
```
Comment (juanignaciosl) [inline app/models/carto/template.rb]: IMHO this also belongs to table.

File: app/models/carto/visualization.rb [hunk 17]
```diff
@@ -324,4 +328,9 @@ def configuration
     Cartodb.config
   end
 
+  def get_templates_using_vis()
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Why not mapping the relationship?

File: app/controllers/carto/api/visualizations_controller.rb (hunk 11 above)
Comment (Kartones) [inline app/controllers/carto/api/visualizations_controller.rb]: 1) Yep
2) Could, but load visualization does way more things, from allowing names to checkinv user has visiblity, while templates requirements are organization visibility, not especifically "current user visibility", so I went for a clear and simpler load
3) I also initially thought about it, but wasn't sure and @ethervoid said it could be more misleading. Also while it uses templates it is true that uses them *regarding the parent visualization*, so I opted to add it to the viz controller.

I'm really not sure about 3), because I see benefits and possible extra efforts searching for things in the future in both approaches. If we keep all "regarding a visualization" in the same controller is easier to find even if the controller gets fat, if we split it is a myriad of small controllers and harder to find at a quick glance.

File: db/migrate/20150731082648_create_templates.rb [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    Rails::Sequel.connection.run(%Q{
+      ALTER TABLE "templates"
+        ADD CONSTRAINT  organization_id_fkey
+        FOREIGN KEY (organization_id)
+        REFERENCES organizations(id)
+        ON DELETE CASCADE
+      })
+
+    Rails::Sequel.connection.run(%Q{
+      CREATE INDEX organization_id_idx ON templates(organization_id)
```
Comment (juanignaciosl) [inline db/migrate/20150731082648_create_templates.rb]: Maybe an index on `source_visualization_id` is also useful, isn't it?

File: app/controllers/carto/api/template_presenter.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/template_presenter.rb]: Ok, as you wish. I'll try to remember it from now on.

File: app/controllers/carto/api/templates_controller.rb (hunk 8 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: Ok. Just setters, you're totally right about its simplicity.

File: app/models/carto/template.rb (hunk 13 above)
Comment (Kartones) [inline app/models/carto/template.rb]: which one again, the required_tables? ok, I'll document it (thought was clear but yep, better)

File: app/controllers/carto/api/templates_controller.rb (hunk 10 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/templates_controller.rb]: Ok. I use to doubt because of security/privacy concerns, but it's true. I should attach to standards.

File: app/controllers/carto/api/visualizations_controller.rb (hunk 11 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: 1) Ok.
2) Ok.
3) I think the right criteria about controllers is "if it returns X (as its first-level-value) it should be in XController". I think it's clear and avoids code duplication from the very beginning. It just adds some filtering alternatives.

File: app/models/carto/template.rb (hunk 15 above)
Comment (Kartones) [inline app/models/carto/template.rb]: To make sure the visualization not only exists (whih DB FK checks) but that it belongs to your organization... requirement for the templates

File: app/models/carto/template.rb (hunk 14 above)
Comment (Kartones) [inline app/models/carto/template.rb]: makes sense, yep. will change it

File: app/models/carto/template.rb (hunk 12 above)
Comment (juanignaciosl) [inline app/models/carto/template.rb]: I meant that line, `def ==`. According to docs, [ActiveRecord already delegates comparison through id](http://apidock.com/rails/v3.2.13/ActiveResource/Base/%3D%3D). I think it was needed for old models. I see it's at `Carto::Visualization`, if I'm right it should be deleted as well.

File: app/models/carto/template.rb (hunk 13 above)
Comment (juanignaciosl) [inline app/models/carto/template.rb]: I meant the reader and writers, which are a workaround for `required_tables_list`, since if extension worked it wouldn't be necessary. It's clear for me today, but not for my future me :-)

File: app/models/carto/template.rb (hunk 15 above)
Comment (juanignaciosl) [inline app/models/carto/template.rb]: Oh, sure! Sorry!

File: db/migrate/20150731082648_create_templates.rb (hunk 18 above)
Comment (Kartones) [inline db/migrate/20150731082648_create_templates.rb]: indeed, adding it

File: app/models/carto/template.rb (hunk 12 above)
Comment (Kartones) [inline app/models/carto/template.rb]: ahh, yeap, it was because at old models was required, removing it then.

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5605/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5606/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5609/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5617/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5623/))

**Comment by juanignaciosl** [comment]
:+1:

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5628/))

**Comment by Kartones** [comment]
At a staging, all looks fine:
```
                                  Table "public.templates"
         Column          |            Type             |              Modifiers              
-------------------------+-----------------------------+-------------------------------------
 id                      | uuid                        | not null default uuid_generate_v4()
 source_visualization_id | uuid                        | not null
 title                   | text                        | not null
 description             | text                        | not null default ''::text
 min_supported_version   | text                        | not null default ''::text
 max_supported_version   | text                        | not null default ''::text
 code                    | text                        | not null default ''::text
 organization_id         | uuid                        | not null
 required_tables_list    | text                        | not null default ''::text
 created_at              | timestamp without time zone | not null default now()
Indexes:
    "templates_pkey" PRIMARY KEY, btree (id)
    "organization_id_idx" btree (organization_id)
    "source_visualization_id_idx" btree (source_visualization_id)
Foreign-key constraints:
    "organization_id_fkey_cascade_del" FOREIGN KEY (organization_id) REFERENCES organizations(id) ON DELETE CASCADE
    "source_visualization_id_fkey_cascade_del" FOREIGN KEY (source_visualization_id) REFERENCES visualizations(id) ON DELETE CASCADE
```
So, for now disabling the offending test until we take down current CI system.

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/5638/))
